File: backend/sdk/python/beacon/tracer.py
import os
import uuid
import time
import json
import requests
from typing import Optional, Dict, Any, Callable
from functools import wraps

class BeaconTracer:
    def __init__(self, api_url: str = "http://localhost:8000", agent_id: str = "default"):
        self.api_url = api_url.rstrip("/")
        self.agent_id = agent_id
        self._active_trace: Optional[str] = None
        self._steps: list = []

    def start_trace(self, metadata: Optional[Dict[str, Any]] = None, parent_trace_id: Optional[str] = None) -> str:
        trace_id = str(uuid.uuid4())
        self._active_trace = trace_id
        self._steps = []

        try:
            payload = {"id": trace_id, "agent_id": self.agent_id, "metadata": metadata}
            if parent_trace_id:
                payload["parent_trace_id"] = parent_trace_id
            requests.post(
                f"{self.api_url}/api/v1/traces/",
                json=payload,
                timeout=2
            )
        except Exception:
            pass  # Fail silently — observability must not break the agent

        return trace_id

    def log_step(self, step_type: str, tool_name: str = None, model: str = None,
                 input_tokens: int = 0, output_tokens: int = 0, cost_usd: float = 0.0,
                 latency_ms: int = 0, status: str = "ok", error: str = None,
                 tool_status: str = "ok", tool_response_code: int = None, tool_error: str = None,
                 retry_count: int = 0, retry_history: list = None):
        self._steps.append({
            "step_number": len(self._steps) + 1,
            "type": step_type,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S"),
            "duration_ms": latency_ms,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
            "cost_usd": cost_usd,
            "tool_name": tool_name,
            "model": model,
            "latency_ms": latency_ms,
            "status": status,
            "error": error,
            "tool_status": tool_status,
            "tool_response_code": tool_response_code,
            "tool_error": tool_error,
            "retry_count": retry_count,
            "retry_history": retry_history
        })

    def end_trace(self, status: str = "success", error_message: str = None, 
                  first_failure_step_number: int = None, first_failure_reason: str = None,
                  retry_count: int = None):
        if not self._active_trace:
            return

        total_tokens = sum(s["input_tokens"] + s["output_tokens"] for s in self._steps)
        total_cost = sum(s["cost_usd"] for s in self._steps)
        
        if retry_count is None:
            retry_count = sum(s.get("retry_count", 0) for s in self._steps)

        try:
            payload = {
                "status": status,
                "ended_at": time.strftime("%Y-%m-%dT%H:%M:%S"),
                "total_tokens": total_tokens,
                "prompt_tokens": sum(s["input_tokens"] for s in self._steps),
                "completion_tokens": sum(s["output_tokens"] for s in self._steps),
                "cost_usd": total_cost,
                "steps": self._steps,
                "error_message": error_message,
                "retry_count": retry_count
            }
            if first_failure_step_number is not None:
                payload["first_failure_step_number"] = first_failure_step_number
            if first_failure_reason:
                payload["first_failure_reason"] = first_failure_reason
            
            resp = requests.patch(
                f"{self.api_url}/api/v1/traces/{self._active_trace}",
                json=payload,
                timeout=2
            )
            if resp.status_code != 200:
                print(f"Error updating trace: {resp.status_code} - {resp.text}")
        except Exception as e:
            print(f"Exception in end_trace: {e}")

        self._active_trace = None
        self._steps = []

    def trace(self, func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            self.start_trace(metadata={"function": func.__name__})
            start = time.time()
            try:
                result = func(*args, **kwargs)
                self.end_trace(status="success")
                return result
            except Exception as e:
                self.end_trace(status="failure", error_message=str(e))
                raise
        return wrapper

File: backend/sdk/typescript/src/tracer.ts
import { v4 as uuidv4 } from 'uuid';

export interface Step {
  step_number: number;
  type: string;
  timestamp: string;
  duration_ms: number;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number;
  tool_name?: string;
  model?: string;
  latency_ms: number;
  status: string;
  error?: string;
  tool_status?: string;
  tool_response_code?: number;
  tool_error?: string;
  retry_count?: number;
  retry_history?: any[];
}

export interface TraceMetadata {
  [key: string]: any;
}

export class BeaconTracer {
  private apiUrl: string;
  private agentId: string;
  private activeTrace: string | null = null;
  private steps: Step[] = [];

  constructor(apiUrl: string = 'http://localhost:8000', agentId: string = 'default') {
    this.apiUrl = apiUrl.replace(/\/$/, '');
    this.agentId = agentId;
  }

  async startTrace(metadata?: TraceMetadata, parentTraceId?: string): Promise<string> {
    const traceId = uuidv4();
    this.activeTrace = traceId;
    this.steps = [];

    try {
      const payload: any = { id: traceId, agent_id: this.agentId, metadata };
      if (parentTraceId) payload.parent_trace_id = parentTraceId;
      
      await fetch(`${this.apiUrl}/api/v1/traces/`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
      });
    } catch {
      // Fail silently
    }

    return traceId;
  }

  logStep(params: {
    stepType: string;
    toolName?: string;
    model?: string;
    inputTokens?: number;
    outputTokens?: number;
    costUsd?: number;
    latencyMs?: number;
    status?: string;
    error?: string;
    toolStatus?: string;
    toolResponseCode?: number;
    toolError?: string;
    retryCount?: number;
    retryHistory?: any[];
  }): void {
    this.steps.push({
      step_number: this.steps.length + 1,
      type: params.stepType,
      timestamp: new Date().toISOString(),
      duration_ms: params.latencyMs || 0,
      input_tokens: params.inputTokens || 0,
      output_tokens: params.outputTokens || 0,
      cost_usd: params.costUsd || 0,
      tool_name: params.toolName,
      model: params.model,
      latency_ms: params.latencyMs || 0,
      status: params.status || 'ok',
      error: params.error,
      tool_status: params.toolStatus || 'ok',
      tool_response_code: params.toolResponseCode,
      tool_error: params.toolError,
      retry_count: params.retryCount || 0,
      retry_history: params.retryHistory,
    });
  }

  async endTrace(
    status: string = 'success',
    errorMessage?: string,
    firstFailureStepNumber?: number,
    firstFailureReason?: string,
    retryCount?: number
  ): Promise<void> {
    if (!this.activeTrace) return;

    const totalTokens = this.steps.reduce((s, step) => s + step.input_tokens + step.output_tokens, 0);
    const totalCost = this.steps.reduce((s, step) => s + step.cost_usd, 0);
    const calculatedRetryCount = retryCount ?? this.steps.reduce((s, step) => s + (step.retry_count || 0), 0);

    try {
      const payload: any = {
        status,
        ended_at: new Date().toISOString(),
        total_tokens: totalTokens,
        prompt_tokens: this.steps.reduce((s, step) => s + step.input_tokens, 0),
        completion_tokens: this.steps.reduce((s, step) => s + step.output_tokens, 0),
        cost_usd: totalCost,
        steps: this.steps,
        error_message: errorMessage,
        retry_count: calculatedRetryCount,
      };
      
      if (firstFailureStepNumber !== undefined) payload.first_failure_step_number = firstFailureStepNumber;
      if (firstFailureReason) payload.first_failure_reason = firstFailureReason;

      await fetch(`${this.apiUrl}/api/v1/traces/${this.activeTrace}`, {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
      });
    } catch {
      // Fail silently
    }

    this.activeTrace = null;
    this.steps = [];
  }

  wrap<T extends (...args: any[]) => any>(fn: T, metadata?: TraceMetadata): T {
    const self = this;
    return (async (...args: any[]) => {
      await self.startTrace(metadata);
      const start = Date.now();
      try {
        const result = await fn(...args);
        await self.endTrace('success');
        return result;
      } catch (error) {
        await self.endTrace('failure', error instanceof Error ? error.message : String(error));
        throw error;
      }
    }) as T;
  }
}
